### PR TITLE
ARROW-92: Arrow to Parquet Schema conversion

### PR DESCRIPTION
**Pull request by xhochy**
My current WIP state. To make the actual schema conversion complete, we probably need the physical structure too as Arrow schemas only care about logical types whereas Parquet schema is about logical and physical types.

### CODE REVIEW

**Comment by wesm** [comment]
We'll have to make some decisions about type mappings. For example:

- `arrow::StringType` becomes `BYTE_ARRAY` with `UTF8` annotation
- `arrow::BinaryType` (needs to be implemented) becomes BYTE_ARRAY with no ConvertedType
- `arrow::CharType` (if ever used, we can skip it for now) becomes FIXED_LEN_BYTE_ARRAY

For List types, we should use the 3-level array encoding as described here https://github.com/apache/parquet-cpp/blob/master/src/parquet/schema/types.h#L49

What other parts do you think are underspecified?

**Comment by xhochy** [comment]
For `Decimal` we need to decide if `use the smallest possible physical type` is the correct strategy.

**Comment by wesm** [comment]
I see. For decimals, I agree we either need multiple Arrow types, or add metadata indicating the physical storage type to the DecimalType. I would say it's better to make this explicit in the Arrow data type, let me know what you think

**Comment by xhochy** [comment]
Probably a simple `storage_type` field could be enough for the DecimalType. As this probably needs to go into the spec, I made separate issues for this https://issues.apache.org/jira/browse/ARROW-183 and https://issues.apache.org/jira/browse/ARROW-184

**Comment by xhochy** [comment]
PR is now in state for a minimal schema conversion basis for Pandas<->Parquet.

File: cpp/src/arrow/parquet/schema.cc [hunk 1]
```diff
@@ -182,6 +185,126 @@ Status FromParquetSchema(
...
+      break;
+    case Type::FLOAT:
+      type = ParquetType::FLOAT;
+      break;
+    case Type::DOUBLE:
+      type = ParquetType::DOUBLE;
+      break;
+    case Type::CHAR:
+      type = ParquetType::FIXED_LEN_BYTE_ARRAY;
+      logical_type = LogicalType::UTF8;
```
Comment (wesm) [inline cpp/src/arrow/parquet/schema.cc]: Aside: we'll need to visit the string encoding question, as logical unicode characters won't map nearly onto a `char(n)` type

File: cpp/src/arrow/parquet/schema.cc [hunk 2]
```diff
@@ -182,6 +185,126 @@ Status FromParquetSchema(
...
+Status ToParquetSchema(
+    const Schema* arrow_schema, std::shared_ptr<::parquet::SchemaDescriptor>* out) {
+  std::vector<NodePtr> nodes(arrow_schema->num_fields());
+  for (int i = 0; i < arrow_schema->num_fields(); i++) {
+    RETURN_NOT_OK(FieldToNode(arrow_schema->field(i), &nodes[i]));
+  }
+
+  NodePtr schema = GroupNode::Make("schema", Repetition::REPEATED, nodes);
+  *out = std::make_shared<::parquet::SchemaDescriptor>();
+  (*out)->Init(schema);
```
Comment (wesm) [inline cpp/src/arrow/parquet/schema.cc]: `Init` can throw an exception. We will need some macro to convert `ParquetException` into an `arrow::Status`

**Comment by wesm** [comment]
This looks good outside the exception handling question

File: cpp/src/arrow/parquet/schema.cc (hunk 2 above)
Comment (xhochy) [inline cpp/src/arrow/parquet/schema.cc]: As we cannot distinguish why a `ParquetException` was thrown, I just mapped all to `Status::Invalid`.

Comment (wesm) [inline cpp/src/arrow/parquet/schema.cc]: We'll probably need to add Status type for "thirdparty exceptions".

**Comment by wesm** [comment]
+1, thank you

**Pull request closed without merging**